staging/src/k8s.io/apiserver/pkg/util/flowcontrol/watch_tracker.go

```diff
@@ -0,0 +1,152 @@
+	RegisterWatch(requestInfo *request.RequestInfo) ForgetWatchFunc
+
+	// GetInterestedWatchCount returns the number of watches that are
+	// potentially interested in a request with a given RequestInfo
+	// for the purpose of estimating cost of that request.
+	GetInterestedWatchCount(requestInfo *request.RequestInfo) int
+}
+
+// watchTracker tracks the number of watches in the system for
+// the purpose of estimating the cost of incoming mutating requests.
+type watchTracker struct {
+	lock sync.Mutex
+
+	watchCount map[watchIdentifier]int
+}
+
+func NewWatchTracker() WatchTracker {
+	return &watchTracker{
+		watchCount: make(map[watchIdentifier]int),
+	}
+}
+
+func getResourceType(requestInfo *request.RequestInfo) string {
+	// resourceType has to include APIGroup to distinguish same type names across groups.
+	// APIVersion should NOT be included as those effectively represent the same objects.
+	return fmt.Sprintf("%s/%s", requestInfo.APIGroup, requestInfo.Resource)
+}
+
+// RegisterWatch implements WatchTracker interface.
+func (w *watchTracker) RegisterWatch(requestInfo *request.RequestInfo) ForgetWatchFunc {
+	if requestInfo == nil || requestInfo.Verb != "watch" {
+		return nil
+	}
+
+	identifier := &watchIdentifier{
+		resourceType: getResourceType(requestInfo),
+		namespace:    requestInfo.Namespace,
+		name:         requestInfo.Name,
+	}
+
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.watchCount[*identifier]++
+	return w.forgetWatch(identifier)
+}
+
+func (w *watchTracker) forgetWatch(identifier *watchIdentifier) ForgetWatchFunc {
+	return func() {
+		w.lock.Lock()
+		defer w.lock.Unlock()
+
+		w.watchCount[*identifier]--
+		if w.watchCount[*identifier] == 0 {
+			delete(w.watchCount, *identifier)
+		}
+	}
+}
+
+// GetInterestedWatchCount implements WatchTracker interface.
+//
+// TODO(wojtek-t): As of now, requestInfo for object creation (POST) doesn't
+//  contain the Name field set. Figure out if we can somehow get it for the
+//  more accurate cost estimation.
+//
+// TODO(wojtek-t): Figure out how to approach DELETECOLLECTION calls.
+func (w *watchTracker) GetInterestedWatchCount(requestInfo *request.RequestInfo) int {
+	if requestInfo == nil || readOnlyVerbs.Has(requestInfo.Verb) {
+		return 0
+	}
+
+	result := 0
+	// The watches that we're interested in include:
+	// - watches for all objects of a resource type (no namespace and name specified)
+	// - watches for all objects of a resource type in the same namespace (no name specified)
+	// - watched interested in this particular object
+	identifier := &watchIdentifier{
+		resourceType: getResourceType(requestInfo),
+	}
+
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	result += w.watchCount[*identifier]
+
+	if requestInfo.Namespace != "" {
+		identifier.namespace = requestInfo.Namespace
+		result += w.watchCount[*identifier]
+	}
+
+	if requestInfo.Name != "" {
+		identifier.name = requestInfo.Name
+		result += w.watchCount[*identifier]
+	}
```

Comment:
This should come before the corresponding code for namespace.

Reply:
No - it's correct as is.

The structure here is:
- watches for all objects of a given type [L139]
- watches for all objects of a given type from a given namespace [L141-144]
- watches for only this particular object [namespace and name specified] [L146-149]

Reply:
oh, right.